src/nc_ssl.c

```diff
@@ -290,6 +290,50 @@ nc_ssl_read(SSL *ssl, void *buf, size_t num) {
         }
     }
 
+    log_debug(LOG_VERB, "read %d initial bytes", bytes_read);
+
+    // OpenSSL reads data in record size blocks (16kb for SSLv3/TLSv1, see
+    // SSL_read man page for details). Sometimes, SSL_read returns data but
+    // there is still more data available from the next block. Keep reading
+    // until there is actually no more data.
+    //
+    // Don't use SSL_pending since it does not reliably work.
+    for (;;) {
```

Comment:
Are we guaranteed to break out of this?

What happens if we have data to read but no space to read it in buf?

Reply:
I believe we will break out every time, but I can make it easier to read. I'll do that.

We won't call `SSL_read` if `bytes_remaining` is zero (once I make the change from your first comment). So we won't read data beyond the space we have to store it. (The contract of this function is that `buf` has space for `num` bytes.)